src/tools/r2rdump/R2RReader.cs

```diff
@@ -65,130 +71,177 @@ public unsafe R2RReader(string filename)
+                do
+                {
+                    int startRva = NativeReader.ReadInt32(Image, ref curOffset);
+                    int endRva = -1;
+                    if (Machine == Machine.Amd64)
                     {
-                        int runtimeFunctionId = method.EntryPointRuntimeFunctionId;
-                        if (runtimeFunctionId == -1)
-                            continue;
-                        curOffset = runtimeFunctionOffset + runtimeFunctionId * runtimeFunctionSize;
-                        do
-                        {
-                            int startRva = NativeReader.ReadInt32(Image, ref curOffset);
-                            int endRva = -1;
-                            if (Machine == Machine.Amd64)
-                            {
-                                endRva = NativeReader.ReadInt32(Image, ref curOffset);
-                            }
-                            int unwindRva = NativeReader.ReadInt32(Image, ref curOffset);
-
-                            method.RuntimeFunctions.Add(new RuntimeFunction(runtimeFunctionId, startRva, endRva, unwindRva, method));
-                            runtimeFunctionId++;
-                        }
-                        while (runtimeFunctionId < nRuntimeFunctions && !isEntryPoint[runtimeFunctionId]);
+                        endRva = NativeReader.ReadInt32(Image, ref curOffset);
                     }
+                    int unwindRva = NativeReader.ReadInt32(Image, ref curOffset);
+
+                    method.RuntimeFunctions.Add(new RuntimeFunction(runtimeFunctionId, startRva, endRva, unwindRva, method));
+                    runtimeFunctionId++;
                 }
+                while (runtimeFunctionId < isEntryPoint.Length && !isEntryPoint[runtimeFunctionId]);
+            }
+        }
+
+        private void ParseAvailableTypes()
+        {
+            R2RSection instMethodEntryPointSection = R2RHeader.Sections[R2RSection.SectionType.READYTORUN_SECTION_AVAILABLE_TYPES];
+            int instMethodEntryPointsOffset = GetOffset(instMethodEntryPointSection.RelativeVirtualAddress);
+            NativeParser parser = new NativeParser(Image, (uint)instMethodEntryPointsOffset);
+            NativeHashtable instMethodEntryPoints = new NativeHashtable(Image, parser);
+            NativeHashtable.AllEntriesEnumerator allEntriesEnum = instMethodEntryPoints.EnumerateAllEntries();
+            NativeParser curParser = allEntriesEnum.GetNext();
+            while (!curParser.IsNull())
+            {
+                uint rid = curParser.GetUnsigned();
+                rid = rid >> 1;
+                TypeDefinitionHandle typeDefHandle = MetadataTokens.TypeDefinitionHandle((int)rid);
+
+                TypeDefinition typeDef = _mdReader.GetTypeDefinition(typeDefHandle);
+                Tuple<string, string> pair = new Tuple<string, string>(_mdReader.GetString(typeDef.Namespace), _mdReader.GetString(typeDef.Name));
```

Comment:
Two comments:

1) How come you don't combine the namespace and type name right here? Do you plan to use them separately later on?
2) You might be able to use the _mdReader for when you're getting the name of methods.

Reply:
I don't use them separately right now since it's only printing it, but would it be useful to have them separate?